### PR TITLE
rule: role-has-required-aria-props

### PR DESCRIPTION
**Pull request by ematipico**
First draft implementation of the [upstream](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/master/docs/rules/role-has-required-aria-props.md) rule.

I'd like to have some feedback about the first implementation. Because of the fact that `aria-*` attributes and roles, I thought to put all these utilities inside its own package, let me know if I need to move it somewhere else.

I took the mappings from here https://github.com/A11yance/aria-query/blob/master/flow/aria.js and adapted them to TypeScript

### CODE REVIEW

File: packages/@romejs/aria/roles/Role.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    constructor(role: ARIARoleDefinition) {
+        this.role = role;
+    }
+
+    getRequiredProps() {
+        return this.role.requiredProps;
+    }
+
+    getRequiredPropsAsKeys() {
+        const keys = <T extends Object>(obj: T): (keyof T)[] => Object.keys(this.getRequiredProps()) as any;
```
Comment (ematipico) [inline packages/@romejs/aria/roles/Role.ts:15]: By design `Object.keys` won't work with the current structure of the types.

Check here [this](https://www.typescriptlang.org/play/index.html?ssl=13&ssc=3&pln=1&pc=1#code/FASwdgLgpgTgZgQwMZQAQDED2nUG9iqoBGCMA-AFyoDOEM4A5gNzAC+wHSmYtqCVWHAF48BYqSoByAKyS2HAPJEAVlCQQAdAGsoAT2oAKBAEoNcTDACiyABYGdu1EIB8owlx4RUANwQAbAFc0EQQAbQcAXRZWYyA)

File: packages/@romejs/compiler/lint/rules/jsx-a11y/roleHasRequiredAriaProps.test.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+test(
+	"role has required aria props",
+	async (t) => {
+		await testLintMultiple(
+			t,
+			[
+				// INVALID
+				"<span role='checkbox'></span>",
+				// VALID
+				"<span role='checkbox' aria-checked='true'></span>",
```
Comment (ematipico) [inline packages/@romejs/compiler/lint/rules/jsx-a11y/roleHasRequiredAriaProps.test.ts:13]: For now I wanted to make a test against the role `checkbox` just to make a sense of the implementation.

**Comment by ematipico** [comment]
Thank you for your feedback! Sure I can move stuff around, it's not a problem! :) 
Sorry I completely missed that halper file

**Comment by ematipico** [comment]
@bitpshr  I followed your suggestion and moved stuff around. Now, for now the code implemented as it is just checks the presence of the required `aria-*` attributes. I was wondering of this rule should check also the *value* of the property?

With this rule, a value `aria-checked="dog"` is accepted.

**Comment by ematipico** [comment]
Added to the rules all the required props for certain aria roles!

File: packages/@romejs/compiler/lint/rules/ariaHelpers.ts [hunk 3]
```diff
@@ -374,3 +374,98 @@ export const ariaPropsMap: MapOfARIAPropertyDefinitions = new Map([
 		},
 	],
 ]);
+
+export type ARIARoleDefinition = {
+	/* aria-* properties and states allowed on this role. */
+	props: Array<ARIAProperty>;
+	/* aria-* properties and states required on this role. */
+	requiredProps: Array<ARIAProperty>;
```
Comment (bitpshr) [inline packages/@romejs/compiler/lint/rules/ariaHelpers.ts:382]:
```suggestion
	props: Array<ARIAProperty>;
	requiredProps: Array<ARIAProperty>;
```

File: packages/@romejs/compiler/lint/rules/jsx-a11y/roleHasRequiredAriaProps.test.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
+import {test} from "rome";
+import {testLintMultiple} from "../testHelpers";
+
+test(
+	"role has required aria props",
```
Comment (bitpshr) [inline packages/@romejs/compiler/lint/rules/jsx-a11y/roleHasRequiredAriaProps.test.ts:5]:
```suggestion
	"jsx a11y role has required aria props",
```

File: packages/@romejs/compiler/lint/rules/jsx-a11y/roleHasRequiredAriaProps.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+import {descriptions} from "@romejs/diagnostics";
+import {
+	getJSXAttribute,
+	hasJSXAttribute,
+	isJSXElement,
+} from "@romejs/js-ast-utils";
+import {roles} from "@romejs/compiler/lint/rules/ariaHelpers";
+
+export default {
+	name: "roleHasRequiredAriaProps",
```
Comment (bitpshr) [inline packages/@romejs/compiler/lint/rules/jsx-a11y/roleHasRequiredAriaProps.ts:11]:
```suggestion
	name: "jsxA11YRoleHasRequiredAriaProps",
```

File: packages/@romejs/diagnostics/descriptions/lint.ts [hunk 6]
```diff
@@ -296,6 +296,20 @@ export const lint = createDiagnosticsCategory({
 			},
 		],
 	},
+	JSX_A11Y_ROLE_HAS_REQUIRED_ARIA_PROPS: (
+		roleName: string,
+		missingAttributes: Array<string>,
+	) => ({
+		category: "lint/jsx-a11y/roleHasRequiredAriaProps",
+		message: `The element with role ${roleName} doesn't have the mandatory aria-* attributes`,
```
Comment (bitpshr) [inline packages/@romejs/diagnostics/descriptions/lint.ts:304]:
```suggestion
		message: `The element with the <emphasis>${roleName}</emphasis> ARIA role does not have the required ARIA attributes.`
```

**Pull request merged**